### PR TITLE
TIP-4.6 (Upgradeable NFT) standard documentation

### PR DESCRIPTION
**Pull request by rustvalve**

### CODE REVIEW

File: src/standard/TIP-4/6.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+### TIP4_6Collection.platformCodeInfo()
+```solidity
+function platformCodeInfo() external view responsible returns (uint256 codeHash, uint16 codeDepth);
+```
+
+Returns the hash and depth of the NftPlatform code. These can be used to calculate the NFT address in third-party contracts. The hash is taken from the NftPlatform code containing the `collection`(`address`) salt.
+
+### TIP4_6Collection.gasUpgradeValue()
+```solidity
+function gasUpgradeValue() external view responsible returns (uint128 fixedValue, uint128 dynamicValue);
```
Comment (rualekseev) [inline src/standard/TIP-4/6.md:72]: What the different between *fixedValue* and *dynamicValue*
How to use them?

File: src/standard/TIP-4/6.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+Returns the amount of gas required to upgrade the NFT.
+
+### Events
+```solidity
+event UpgradeNftRequested(uint32 oldVersion, uint32 newVersion, address nft, address initiator);
+event NftCodeUpdated(uint32 oldVersion, uint32 newVersion, uint256 oldCodeHash, uint256 newCodeHash);
+```
+**UpgradeNftRequested**
+
+You must emit it when an update is requested from the NFT and upgrade is available (the NFT version at the time of the upgrade request is less than the current NFT version in the collection).
```
Comment (rualekseev) [inline src/standard/TIP-4/6.md:84]: Do I MUST emit this event only if NFT updated successfully?

If NFT request update, but don't updated, what I need to do?

File: src/standard/TIP-4/6.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+### TIP4_6Collection.nftVersion()
+```solidity
+function nftVersion() external view responsible returns (uint32);
+```
+
+Returns the current version of the NFT. The Collection stores only the latest version, update history is stored in events.
+
+### TIP4_6Collection.platformCode()
+```solidity
+function nftVersion() external view responsible returns (uint32);
```
Comment (rualekseev) [inline src/standard/TIP-4/6.md]: misstypo
you must use 
function platformCode() external view responsible returns (TvmCell);

File: src/standard/TIP-4/6.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+```
+
+Returns the NftPlatform code containing the `collection`(`address`) salt. The NftPlatform code is non-upgradeable.
+
+### TIP4_6Collection.platformCodeInfo()
+```solidity
+function platformCodeInfo() external view responsible returns (uint256 codeHash, uint16 codeDepth);
+```
+
+Returns the hash and depth of the NftPlatform code. These can be used to calculate the NFT address in third-party contracts. The hash is taken from the NftPlatform code containing the `collection`(`address`) salt.
```
Comment (rualekseev) [inline src/standard/TIP-4/6.md:68]: If you explain more (how to calculate NFT address in third-party contracts to use codeHash and codeDepth

File: src/standard/TIP-4/6.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+```
+
+### TIP4_6Collection.nftVersion()
+```solidity
+function nftVersion() external view responsible returns (uint32);
+```
+
+Returns the current version of the NFT. The Collection stores only the latest version, update history is stored in events.
+
+### TIP4_6Collection.platformCode()
```
Comment (rualekseev) [inline src/standard/TIP-4/6.md:56]: do we really need this function in the standard?

File: src/standard/TIP-4/6.md (hunk 1 above)
Comment (rustvalve) [reply src/standard/TIP-4/6.md:72]: This is necessary for working in different networks. For example in Venom the gas is different, and maybe someday there will be a switch to dynamic gas.
The fixed part includes what goes to the initial balance of the contract, and its compensation, interaction with external contracts that require fixed gas.
The dynamic part includes gasfee and frwfee.
You can use only fixed part and return 0 in dynamic

File: src/standard/TIP-4/6.md (hunk 2 above)
Comment (rustvalve) [reply src/standard/TIP-4/6.md:84]: Yes, you must only emit this event if an update is available. In that case, the collection emits the event and calls the NFT function to update it.
If there is no update, the remaining gas is returned to the owner and the process aborts

File: src/standard/TIP-4/6.md (hunk 3 above)
Comment (rustvalve) [reply src/standard/TIP-4/6.md]: Fixed

File: src/standard/TIP-4/6.md (hunk 5 above)
Comment (rustvalve) [reply src/standard/TIP-4/6.md:56]: Why not? We already have platform code stored in Collection, so if we need to know the code, this function may be useful

**Review by ilyar — CHANGES_REQUESTED** [review]
You need to add a link to the reference implementation

**Comment by rustvalve** [comment]
> You need to add a link to the reference implementation

Added info to the core-description.md with link https://github.com/grandbazar-io/everscale-tip4.6-contracts

**Review by ilyar — COMMENTED** [review]
LGFM

**Comment by ilyar** [comment]
@everscale-org/docs please review it

File: src/standard/TIP-4/6.md (hunk 5 above)
Comment (nerzh) [reply src/standard/TIP-4/6.md:56]: @rustvalve let's decide, are there cases in which this function is needed right now ? If not, the standard can always be extended in the future if necessary

Comment (rustvalve) [reply src/standard/TIP-4/6.md:56]: > @rustvalve let's decide, are there cases in which this function is needed right now ? If not, the standard can always be extended in the future if necessary

Right now there are no such cases. So basically yes, we can omit this function for now

**Review by rualekseev — APPROVED** [review]

**Review by qwadratic — APPROVED** [review]

**Review by nerzh — APPROVED** [review]

File: src/standard/TIP-4/6.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+**NftUpgraded**
+
+You must emit it after the NFT upgrade.
+
+### Upgrade
+
+The NFT upgrade scenario is as follows:
+
+1. The user (in the general case, the NFT manager) calls the NFT `requestUpgrade` function.
+2. The NFT requests an upgrade from the collection, passing information about itself (`id`, `version`).
```
Comment (vp-mazekine) [inline src/standard/TIP-4/6.md:139]: Which function, precisely?

File: src/standard/TIP-4/6.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+3. The Collection compares the version it holds with the one passed by the NFT.
```
Comment (vp-mazekine) [inline src/standard/TIP-4/6.md:140]: + checks that NFT belongs to this collection

**Review by cryshado — CHANGES_REQUESTED** [review]
Required approval:
- Please remove the `platformCodeInfo` method.
- Please remove the `gasUpgradeValue` method.

Recommendations for the the reference implementation:
- Use type `varUint16` for storing any stuff like "coins".
- The platform's contract code can be made much smaller (in size of bits/refs).
- You shouldn't use `time`, `pubkey`, `expire` ABI header pragmas in contracts, that don't process ext messages in it.

File: src/standard/TIP-4/6.md (hunk 6 above)
Comment (cryshado) [reply src/standard/TIP-4/6.md:139]: > 

It's an internal communication between the contracts in the NFT system, and it shouldn't be specified by the standard

File: src/standard/TIP-4/6.md (hunk 7 above)
Comment (cryshado) [reply src/standard/TIP-4/6.md:140]: > + checks that NFT belongs to this collection

This is the basic behavior of NFT contracts, and must not be specified by this standard

**Review by ilyar — APPROVED** [review]

**Review by ilyar — APPROVED** [review]

**Pull request merged**